database/mongodb/mongodb.go

```diff
@@ -184,10 +283,99 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
-func (m *Mongo) Lock() error {
+func (m *Mongo) ensureLockTable() error {
+	indexes := m.db.Collection(m.config.Locking.CollectionName).Indexes()
+
+	indexOptions := options.Index().SetUnique(true).SetName(LockIndexName)
+	_, err := indexes.CreateOne(context.TODO(), mongo.IndexModel{
+		Options: indexOptions,
+		Keys:    findFilter{Key: -1},
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the MongoDb type.
+func (m *Mongo) ensureVersionTable() (err error) {
+	if err = m.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := m.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if _, _, err = m.Version(); err != nil {
+		return err
+	}
 	return nil
 }
 
+// Utilizes advisory locking on the config.LockingCollection collection
+// This uses a unique index on the `locking_key` field.
+func (m *Mongo) Lock() error {
+	if !m.config.Locking.Enabled {
+		return nil
+	}
+	pid := os.Getpid()
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = fmt.Sprintf("Could not determine hostname. Error: %s", err.Error())
+	}
+
+	newLockObj := lockObj{
+		Key:       lockKeyUniqueValue,
+		Pid:       pid,
+		Hostname:  hostname,
+		CreatedAt: time.Now(),
+	}
+	operation := func() error {
+		timeout, cancelFunc := context.WithTimeout(context.Background(), contextWaitTimeout)
```

Comment:
Oh, is there a way to get the current wait time from the exponential backoff? e.g. this could wait too long during the earlier tries of exponential backoff and not long enough in the later retries

Reply:
We can use ExponentialBackoff.NextBackOff(), I think that should work.

The backoff sleeps the current thread if the result of the insert is an error, it is unaffected by how long the context takes to run. (unless I am not understanding how the context works, which is entirely possible :)  )

Reply:
Oh, I'm sorry, NVM!
Although reading and writing to the collection is a blocking call, these operations should be quite fast and not significantly impact the retry logic. My original incorrect concern was that the reading and writing operations would be long running blocking calls, significantly extending the retry period to the extent that no retries would be attempted.